### PR TITLE
Get rid of build shell scripts and move them to Python

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

- Wheel building on Windows has unexpected behavior when invoked from MSYS2 Bash, as it alters the environment when calling Bazel.

- Bazel is aiming to get rid of the MSYS2 dependency on Windows; it's best to stick with Python.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+import argparse
+import errno
+import itertools
+import os
+import subprocess
+import sys
+import traceback
+
+try: import ctypes  # necessary for Windows
```
Comment (pcmoritz) [inline build.py]: This shouldn't be needed, can we just use builtin python unicode routines?

File: build.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+except ImportError: ctypes = None
+
+pip = None  # Will be lazily imported if needed (lazy because the rest of this script should run even on installations without pip)
+
+StringishTypes = (str, bytes) if sys.version_info[0] >= 3 else (basestring, str, unicode)
```
Comment (pcmoritz) [inline build.py]: This shouldn't be needed

File: build.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    result = 0
+    try:
+        # The following line installs pyarrow from S3, these wheels have been
+        # generated from https://github.com/ray-project/arrow-build from
+        # the commit listed in the command.
+        if pyarrow_install and result == 0:
+            pip_args = ('install', '-q',
+                '--target=' + os.path.join(program_dir, "python", "ray", "pyarrow_files"), 'pyarrow==0.14.0.RAY',
+                '--find-links', "https://s3-us-west-2.amazonaws.com/arrow-wheels/516e15028091b5e287200b5df77d77f72d9a6c9a/index.html")
+            result = pip_main(list(pip_args)) if use_current_python_interpreter else proc_call((env['PYTHON_BIN_PATH'], '-m', 'pip') + tuple(pip_args), check=False)[0]
```
Comment (pcmoritz) [inline build.py]: It is better to call subprocess.Popen (or checked_call) here I think, similar to how it is done in the shell script at the moment.

File: build.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+def get_current_python_cli_executable():
+    result = sys.executable
+    if sys.platform == 'win32':
+        # On Windows, CLI vs. GUI (Windowed) Python have different names; the Window versions end with 'w'
+        (dirname, basename) = os.path.split(result)
+        (name, ext) = os.path.splitext(basename)
+        result = os.path.join(dirname, name.rstrip('w').rstrip('W') + ext)
+    return result
+
+def call_bazel(bazel, *args, **kwargs):  # SINGLE-THREADED CALLS ONLY
```
Comment (pcmoritz) [inline build.py]: We shoudn't have this. If people want to configure bazel, they should do it via the `.bazelrc`.

Comment (mehrdadn) [reply build.py]: Confused, isn't the build script supposed to call bazel?

File: build.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+import traceback
+
+try: import ctypes  # necessary for Windows
+except ImportError: ctypes = None
+
+pip = None  # Will be lazily imported if needed (lazy because the rest of this script should run even on installations without pip)
+
+StringishTypes = (str, bytes) if sys.version_info[0] >= 3 else (basestring, str, unicode)
+
+def pip_main(*args):
```
Comment (pcmoritz) [inline build.py]: Let's get rid of this and call pip via subprocess, in a similar way as it is done in the shell scripts (but with proper handling of whitespaces in paths).

File: build.py (hunk 3 above)
Comment (mehrdadn) [reply build.py]: Any thoughts on the comment I put in `pip_main`? I explained the reason for this choice there.

Comment (mehrdadn) [reply build.py]: Any thoughts on the comment I put in `pip_main`? I explained the reason for this choice there.

File: build.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    if pip is None: import pip
+    main = getattr(pip, 'main', None)
+    if not main:
+        # This is technically unsupported... https://stackoverflow.com/a/50255019
+        # However, calling pip _externally_ for an interpreter that is _already_ running
+        # doesn't seem like any less fragile a scenario!
+        from pip._internal import main
+    return main(*args)
+
+def uniquify(items):
```
Comment (pcmoritz) [inline build.py]: I don't think we need this. If we really do need it, we can do `list(set(items))`

File: build.py (hunk 1 above)
Comment (mehrdadn) [reply build.py]: I'm using this to call a platform API though?

File: build.py (hunk 2 above)
Comment (mehrdadn) [reply build.py]: I needed it below though? What do you mean?

File: build.py [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+
+def search_path_or_default(name, *args, **kwargs):
+    result = search_path(name, *args, **kwargs)
+    if result is None: result = name
+    return result
+
+def is_native_windows_or_msys(sys_platform):  # Does NOT return true for WSL, which is considered "Linux"
+    return sys_platform == 'msys' or sys_platform == 'win32'
+
+def make_bash_command_line(sep, *commands, **kwargs):  # Constructs Bash commands that are safely quoted
```
Comment (pcmoritz) [inline build.py]: We shouldn't need this and instead make sure we are not constructing unsafe commands.

File: build.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        result = os.path.join(dirname, name.rstrip('w').rstrip('W') + ext)
+    return result
+
+def call_bazel(bazel, *args, **kwargs):  # SINGLE-THREADED CALLS ONLY
+    result = errno.EFAULT
+    initial_environment = os.environ.copy()  # Copy our environment BEFORE making changes
+    try:
+        env = kwargs.get('env')
+        if env is None: env = initial_environment.copy()
+        found_shell = env.get('BAZEL_SH') or find_compatible_shell('bash', allow_wsl=False)
```
Comment (edoakes) [inline build.py]: What is this for? Can't you just use `subprocess`?

**Review by edoakes — CHANGES_REQUESTED** [review]
This is a good idea but this script is extremely hard to understand, and maintainability is a top priority. Please see my comments and try to clean it up generally and I will take another look. Also, adding comments that explain the windows-specfic behavior anywhere that it's non-obvious would be great.

**Comment by mehrdadn** [comment]
> This is a good idea but this script is extremely hard to understand, and maintainability is a top priority. Please see my comments and try to clean it up generally and I will take another look. Also, adding comments that explain the windows-specfic behavior anywhere that it's non-obvious would be great.

@edoakes: It actually turns out I [thought](/ray-project/ray/pull/6082/files/003890afdcf58c66ff0bb0a0f3439a78633ab69a#diff-2955d5257f635de1df53f55a171ca5c7R3) this script is supposed to be run directly by users, but @pcmoritz mentioned that it's actually only supposed to be run from `setup.py`, generally via `pip`. That fundamentally changes what I thought the script is be responsible for, and it means I should take out a lot of the things that don't make sense inside a `pip` installation (like handling of system packages). I'll move the relevant parts into `setup.py` and update the commit.

**Comment by mehrdadn** [comment]
@pcmoritz I just moved things to `setup.py`. Could you let me know (or just fill in yourself) what should be in place of the previous `build.sh` calls (see the `TODO`s)?

File: build.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+                for option_string in action.option_strings:
+                    parts.append(str(option_string))
+                parts[-1] += ' ' + str(args_string)
+            result = ', '.join(parts)
+        return result
+
+def main(program, *args):
+    program_name = os.path.basename(program)
+    program_dir = os.path.dirname(os.path.realpath(__file__))
+    argparser = argparse.ArgumentParser(prog=program_name, formatter_class=SingleMetaVarHelpFormatter)
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply build.py]: This is moot now, but the reason for it was in the StackOverflow link that was a few lines above (it's gone now): https://stackoverflow.com/a/23941599

File: python/setup.py [hunk 10]
```diff
@@ -82,18 +91,81 @@
...
+        if (not BUILD_SH) and SYSTEMROOT and os.path.isfile(os.path.join(SYSTEMROOT, 'System32', 'bash.exe')):
+            msg = " ".join([
+                "You appear to have Bash from WSL, which Bazel is not compatible with.",
+                "To avoid potential problems, please explicitly set the %r environment variable for Bazel."
+            ]) % ('BUILD_SH',)
+            raise ValueError(msg)
+
+    assert os.path.isabs(root_dir), "need current working directory as an absolute path"
+
+    # TODO(mehrdadn): Is this directory needed anymore?
```
Comment (pcmoritz) [inline python/setup.py]: This is actually not needed any more :)

File: java/doc/installation.rst [hunk 11]
```diff
@@ -46,7 +46,7 @@ Then we can start building Ray with the following commands:
   ci/travis/install-bazel.sh
 
   # build native components
-  ./build.sh -l java
+  ./build.sh -l java  # TODO(pcmoritz)
```
Comment (pcmoritz) [inline java/doc/installation.rst]: We can create a small build.sh here that calls `RAY_INSTALL_JAVA=1 pip install -e .` in the java case and `pip install -e .` in the python case here to support some people's workflow :)

**Comment by pcmoritz** [comment]
This looks really great, thanks for doing it!

File: python/setup.py [hunk 12]
```diff
@@ -82,18 +91,81 @@
...
+            os.makedirs(path)
+        except OSError as ex:
+            error_due_to_existing_dir = exist_ok and ex.errno == errno.EEXIST and os.path.isdir(path)
+            if not error_due_to_existing_dir:
+                raise
+
+def download(url):
+    return urllib_request.urlopen(url).read()
+
+def bazel_build(root_dir, build_ext=None):  # Keep this as a global function to make life easier during development
```
Comment (edoakes) [inline python/setup.py]: I don't understand this comment out of context. If it's just coding style, remove it. If it's important, explain it more.

File: python/setup.py [hunk 13]
```diff
@@ -82,18 +91,81 @@
     "tune": ["tabulate"],
 }
 
+def is_native_windows_or_msys():  # Does NOT return true for WSL, which is considered "Linux"
```
Comment (edoakes) [inline python/setup.py]: If you're commenting on the behavior of a function, please use docstring format

File: python/setup.py [hunk 14]
```diff
@@ -82,18 +91,81 @@
...
+
+def download(url):
+    return urllib_request.urlopen(url).read()
+
+def bazel_build(root_dir, build_ext=None):  # Keep this as a global function to make life easier during development
+    if is_native_windows_or_msys():
+        BUILD_SH = os.getenv('BUILD_SH')
+        SYSTEMROOT = os.getenv('SYSTEMROOT')
+        if (not BUILD_SH) and SYSTEMROOT and os.path.isfile(os.path.join(SYSTEMROOT, 'System32', 'bash.exe')):
+            msg = " ".join([
```
Comment (edoakes) [inline python/setup.py]: please use `str.format()` syntax instead of `%`

File: python/setup.py [hunk 15]
```diff
@@ -82,18 +91,81 @@
...
+                "You appear to have Bash from WSL, which Bazel is not compatible with.",
+                "To avoid potential problems, please explicitly set the %r environment variable for Bazel."
+            ]) % ('BUILD_SH',)
+            raise ValueError(msg)
+
+    assert os.path.isabs(root_dir), "need current working directory as an absolute path"
```
Comment (edoakes) [inline python/setup.py]: `assert` statements should only be used for internal invariant checking. Raise a proper exception if this is validation

File: python/setup.py [hunk 16]
```diff
@@ -82,18 +91,81 @@
...
+
+    # TODO(mehrdadn): Is this directory needed anymore?
+    makedirs(os.path.join(root_dir, 'build'), exist_ok=True)
+
+    # Note: We are passing in sys.executable so that we use the same
+    # version of Python to build pyarrow inside the build.sh script. Note
+    # that certain flags will not be passed along such as --user or sudo.
+    # TODO(rkn): Fix this.
+    pyarrow_url = 'https://s3-us-west-2.amazonaws.com/arrow-wheels/3a11193d9530fe8ec7fdb98057f853b708f6f6ae/index.html'
```
Comment (edoakes) [inline python/setup.py]: Can we make this a global variable at the top of the file instead of burying it in the code?

Comment (edoakes) [reply python/setup.py]: same with the pyarrow version, pickle5_backport_url, etc

File: python/setup.py [hunk 17]
```diff
@@ -82,18 +91,81 @@
...
+        pickle5_dir = os.path.join(root_dir, 'python', 'ray', 'pickle5_files')
+        work_dir = tempfile.mkdtemp()
+        try:
+            zipfile.ZipFile(io.BytesIO(download(pickle5_backport_url)), 'r').extractall(work_dir)
+            subprocess.check_call([sys.executable, 'setup.py', 'bdist_wheel'], cwd=work_dir)
+            for wheel in glob.glob(os.path.join(work_dir, 'dist', '*.whl')):
+                zipfile.ZipFile(wheel, 'r').extractall(pickle5_dir)
+        finally:
+            shutil.rmtree(work_dir)
+    bazel_env = dict(os.environ.copy(), **{'PYTHON%d_BIN_PATH' % sys.version_info[0]: sys.executable})
```
Comment (edoakes) [inline python/setup.py]: a lot of python magic in this line. would be nicer if you just do `bazel_env = dict(os.environ.copy()` and then `bazel_env.update({"PYTHON_BIN_PATH": ...})`

File: python/setup.py [hunk 18]
```diff
@@ -82,18 +91,81 @@
...
+    bazel_build_cmd = ['bazel', 'build', '--verbose_failures']
+    bazel_build_cmd.append('//:ray_pkg')
```
Comment (edoakes) [inline python/setup.py]: why not just have this in the above line?

File: python/setup.py [hunk 19]
```diff
@@ -82,18 +91,81 @@
...
-        # version of Python to build pyarrow inside the build.sh script. Note
-        # that certain flags will not be passed along such as --user or sudo.
-        # TODO(rkn): Fix this.
-        command = ["../build.sh", "-p", sys.executable]
-        if os.getenv("RAY_INSTALL_JAVA") == "1":
-            # Also build binaries for Java if the above env variable exists.
-            command += ["-l", "python,java"]
-        subprocess.check_call(command)
+        if is_invalid_windows_platform():
+            # https://github.com/msys2/MINGW-packages/blob/abd06ca92d876b9db05dd65f27d71c4ebe2673a9/mingw-w64-python2/0410-MINGW-build-extensions-with-GCC.patch#L53
```
Comment (edoakes) [inline python/setup.py]: Please add context to this comment. i.e., what the link is pointing to

File: python/setup.py [hunk 20]
```diff
@@ -82,18 +91,81 @@
     "tune": ["tabulate"],
 }
 
+def is_native_windows_or_msys():  # Does NOT return true for WSL, which is considered "Linux"
+    return sys.platform == 'msys' or sys.platform == 'win32'
+
+def is_invalid_windows_platform():
+    return sys.platform == 'msys' or sys.platform == 'win32' and sys.version_string and 'GCC' in sys.version_string
+
+def makedirs(path, exist_ok):
```
Comment (edoakes) [inline python/setup.py]: Please add a docstring comment specifying that this is just for python2 compatibility

**Review by edoakes — DISMISSED** [review]
Nice, much simpler now. Some more style feedback.

File: python/setup.py (hunk 18 above)
Comment (mehrdadn) [reply python/setup.py]: I thought it's more readable visually separating the package names from the option arguments.

File: python/setup.py [hunk 21]
```diff
@@ -20,6 +29,10 @@
 
 # NOTE: The lists below must be kept in sync with ray/BUILD.bazel.
 
+pyarrow_url = "https://s3-us-west-2.amazonaws.com/arrow-wheels/3a11193d9530fe8ec7fdb98057f853b708f6f6ae/index.html"
```
Comment (edoakes) [inline python/setup.py]: please make global constants `ALL_CAPS_SNAKE_CASE`

Comment (mehrdadn) [reply python/setup.py]: @edoakes the other variables below aren't like this though?

File: build.py [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+            pip_args = ('install', '-q',
+                '--target=' + os.path.join(program_dir, "python", "ray", "pyarrow_files"), 'pyarrow==0.14.0.RAY',
+                '--find-links', "https://s3-us-west-2.amazonaws.com/arrow-wheels/516e15028091b5e287200b5df77d77f72d9a6c9a/index.html")
+            result = pip_main(list(pip_args)) if use_current_python_interpreter else proc_call((env['PYTHON_BIN_PATH'], '-m', 'pip') + tuple(pip_args), check=False)[0]
+        if result == 0:
+            result = call_bazel(*command_args)
+    except OSError as ex:
+        result = ex.errno
+    except KeyboardInterrupt:
+        result = errno.EINTR  # or should we return 572 (ERROR_CONTROL_C_EXIT) for Windows?
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply build.py]: I guess it's moot now, but this was to avoid printing an ugly traceback.

File: build.py [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+    use_current_python_interpreter = False
+    if parsed_args.python is not None:
+        env['PYTHON_BIN_PATH'] = parsed_args.python
+    elif 'PYTHON_BIN_PATH' not in env:
+        env.setdefault('PYTHON_BIN_PATH', os.path.normpath(get_current_python_cli_executable()))
+        use_current_python_interpreter = True
+    pyarrow_install = not os.getenv('SKIP_PYARROW_INSTALL') and sys.platform != 'win32'
+    command_args = ['bazel', 'build']
+    for language in parsed_args.language.split(','):
+        if not language: continue
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply build.py]: It was meant to handle adjacent commas gracefully.

File: build.py [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+        if language.startswith('//') or language.startswith('@'):
+            command_args.append(language)
+        elif 'python' in language:
+            ray_build_python = command_args.append('//:ray_pkg')
+        elif 'java' in language:
+            ray_build_java = command_args.append('//java:all')
+        else: raise ValueError("Unrecognized language: " + language)
+    result = 0
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply build.py]: I hate multiple return paths :-) it makes it much harder to understand the control flow.

File: python/setup.py (hunk 21 above)
Comment (edoakes) [reply python/setup.py]: ah, they should be :) but I guess this is fine because it's `setup.py` and not a normal python program...

Comment (mehrdadn) [reply python/setup.py]: Yeah I think it's the setup.py convention.

**Comment by mehrdadn** [comment]
@pcmoritz @edoakes I think this should be ready for merge.

**Comment by edoakes** [comment]
The python tests are unrelated but the linter is failing:
https://travis-ci.com/ray-project/ray/jobs/257550768

You should be able to run `./ci/travis/format.sh` to lint locally. You can also run `./setup_hooks.sh` to set up a github push hook that will run the linter before pushing code to the remote repo.

File: python/setup.py [hunk 25]
```diff
@@ -82,18 +95,113 @@
...
+    if sys.version_info[:2] >= (3, 2):
+        os.makedirs(path, exist_ok=exist_ok)
+    else:
+        try:
+            os.makedirs(path)
+        except OSError as ex:
+            if not (exist_ok and ex.errno == errno.EEXIST and os.path.isdir(path)):
+                raise
+
+def zip_extract_members(zip_file, desired_members, target_dir):
```
Comment (pcmoritz) [inline python/setup.py]: It would really be great to get rid of this function. This functionality is alreayd implement in the `zip` module. Why don't we use it instead of reimplementing it?

Comment (mehrdadn) [reply python/setup.py]: I'm not reimplementing anything -- the functionality doesn't exist in the ZIP module unfortunately. It only allows extracting the entire archive or a single file, not just a subdirectory.

Comment (pcmoritz) [reply python/setup.py]: Can we use the file `https://files.pythonhosted.org/packages/cd/5a/cbdf36134804809d55ffd4c248343bd36680a92b6425885a3fd204d32f7b/pickle5-0.0.9.tar.gz` which doesn't have subdirectories?

Comment (mehrdadn) [reply python/setup.py]: I think the contents aren't the same? But also it seems to me there's still an extra subdirectory like before (as is conventional with tar files)?

Comment (pcmoritz) [reply python/setup.py]:
```
In [1]: import tarfile 
   ...: tf = tarfile.open("pickle5-0.0.9.tar.gz")                                                                                                                                    

In [2]: tf.extractall()
```

works for me

Comment (mehrdadn) [reply python/setup.py]: Are you sure? I'm confused... it seems to create the same directory structure as
```
import zipfile
>>> zf = zipfile.ZipFile("pickle5-backport-5186f9ca4ce55ae530027db196da51e08208a16b.zip", "r")
>>> zf.extractall()
```
In your case I get a `pickle5-0.0.9` folder, in this case I get a `pickle5-backport-5186f9ca4ce55ae530027db196da51e08208a16b` folder, both of which contain `setup.py`.

Comment (pcmoritz) [reply python/setup.py]: Looks like that's indeed the case, but `pickle5-0.0.9` is pretty short, so the danger of path lenght overlow is pretty small and we care less about creating an additional subdirectory :)

Comment (mehrdadn) [reply python/setup.py]: Okay sounds good! Will modify it in a bit.

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
There is a small problem on python 3 currently:
```
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 188, in nti
    s = nts(s, "ascii", "strict")
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 172, in nts
    return s.decode(encoding, errors)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xfd in position 0: ordinal not in range(128)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 2295, in next
    tarinfo = self.tarinfo.fromtarfile(self)
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 1091, in fromtarfile
    obj = cls.frombuf(buf, tarfile.encoding, tarfile.errors)
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 1033, in frombuf
    chksum = nti(buf[148:156])
  File "/home/travis/miniconda/lib/python3.6/tarfile.py", line 191, in nti
    raise InvalidHeaderError("invalid header")
tarfile.InvalidHeaderError: invalid header
```

Probably unicode related, do we need to pass `"b"` somewhere? Also, urllib seems to be broken in the linux wheels environment.

**Comment by mehrdadn** [comment]
Ah silly mistake, I thought I needed to switch to `tarfile.TarFile` from `tarfile.open` in order to use the `fileobj=` parameter... didn't realize the latter had the same ability.

**Comment by mehrdadn** [comment]
@pcmoritz It seems the certificate issue is still there (!)... I have no clue what's going on or how to reproduce...

**Comment by pcmoritz** [comment]
Haha now it's working on Python 3.5 but breaking on 3.6, let me have another look.

**Comment by pcmoritz** [comment]
Should be fixed now, fingers crossed.

**Comment by mehrdadn** [comment]
Seems it's finally ready! :)

**Comment by pcmoritz** [comment]
The Python 3 errors actually look related, it looks like somehow the PR makes it so the old serialization path is being used. I'll investigate into that.

**Comment by mehrdadn** [comment]
Closed this by accident...

**Comment by robertnishihara** [comment]
@mehrdadn this can be closed, right?

**Comment by mehrdadn** [comment]
No, it should be updated (again) and merged.

File: python/setup.py [hunk 26]
```diff
@@ -72,7 +87,7 @@
 optional_ray_files += ray_project_files
 optional_ray_files += ray_dashboard_files
 
-if "RAY_USE_NEW_GCS" in os.environ and os.environ["RAY_USE_NEW_GCS"] == "on":
+if os.getenv("RAY_USE_NEW_GCS") == "on":
```
Comment (edoakes) [inline python/setup.py:90]: This is now making the assumption that `RAY_USE_NEW_GCS` is set as an environment variable - is that safe? Why did we change it?

File: python/setup.py [hunk 27]
```diff
@@ -101,91 +116,151 @@
...
+    try:
+        result = urllib.request.urlopen(url).read()
+    except urllib.error.URLError:
+        # This fallback is necessary on Python 3.5 on macOS due to TLS 1.2.
+        curl_args = ["curl", "-s", "-L", "-f", "-o", "-", url]
+        result = subprocess.check_output(curl_args)
+    return result
+
+
+def download_pickle5(pickle5_dir):
```
Comment (edoakes) [inline python/setup.py:143]: Can you leave a high-level comment about what this is doing? e.g., what is the path it'll be downloaded to (the string manipulation logic is very hard to follow)

**Review by edoakes — COMMENTED** [review]
Looks pretty good. My chief complaint is that I found `download_pickle5` extremely hard to follow - any way to make that a bit more clear at the cost of verbosity?

**Review by edoakes — APPROVED** [review]

File: python/setup.py (hunk 26 above)
Comment (mehrdadn) [reply python/setup.py:90]: Actually `os.getenv` is like `os.environ.get`, it returns `None` if it doesn't exist.

File: python/setup.py (hunk 27 above)
Comment (mehrdadn) [reply python/setup.py:143]: Sure

File: python/setup.py (hunk 26 above)
Comment (edoakes) [reply python/setup.py:90]: Oh, gotcha! Didn't catch that change.

**Comment by mboisson** [comment]
This PR made it impossible to build from source using pip: 
```
Collecting https://github.com/ray-project/ray/archive/ray-1.1.0.tar.gz
  Using cached https://github.com/ray-project/ray/archive/ray-1.1.0.tar.gz
ERROR: https://github.com/ray-project/ray/archive/ray-1.1.0.tar.gz does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
```
